File: 02-tech/01-programming-fundamentals/13-exam/04-softuni-exam-results/SoftuniExamResults.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace _04_softuni_exam_results
{
    class SoftuniExamResults
    {
        static void Main()
        {
            var pointsByParticipant = new SortedDictionary<string, int>();
            var submissionCountByLanguage = new SortedDictionary<string, int>();

            var input = string.Empty;
            while ((input = Console.ReadLine()) != "exam finished")
            {
                var data = input.Split('-');
                var participant = data[0];

                if (data.Length == 3)
                {
                    var language = data[1];
                    var points = int.Parse(data[2]);

                    if (pointsByParticipant.ContainsKey(participant) == false)
                    {
                        pointsByParticipant[participant] = 0;
                    }

                    if (submissionCountByLanguage.ContainsKey(language) == false)
                    {
                        submissionCountByLanguage[language] = 0;
                    }

                    if (pointsByParticipant[participant] < points)
                    {
                        pointsByParticipant[participant] = points;
                    }

                    submissionCountByLanguage[language]++;
                }
                else // Ban
                {
                    pointsByParticipant.Remove(participant);
                }
            }

            Console.WriteLine("Results:");
            foreach (var participant in pointsByParticipant.OrderByDescending(x => x.Value))
            {
                Console.WriteLine($"{participant.Key} | {participant.Value}");
            }

            Console.WriteLine("Submissions:");
            foreach (var language in submissionCountByLanguage.OrderByDescending(x => x.Value))
            {
                Console.WriteLine($"{language.Key} - {language.Value}");
            }
        }
    }
}

File: 02-tech/01-programming-fundamentals/13-exam/04-softuni-exam-results/softuni-exam-results.js
function solve(entries)
{
    const submissionsByLanguage = {};
    const pointsByUser = new Map();
    entries.pop();

    for (let entry of entries)
    {
        const [user, language, currentResult] = entry.split('-');

        if (currentResult) // Entry
        {
            const oldResult = pointsByUser.get(user) || 0;

            pointsByUser.set(user, Math.max(oldResult, +currentResult));

            submissionsByLanguage[language] = submissionsByLanguage[language] + 1 || 1;
        }
        else // Ban
        {
            pointsByUser.delete(user);
        }
    }

    console.log('Results:');
    const sortedPointsByUser = [...pointsByUser.keys()]
        .sort((a, b) =>
        {
            if (pointsByUser.get(a) == pointsByUser.get(b))
            {
                // Ascending by name
                return a.localeCompare(b);
            }
            else
            {
                // Descending by points
                return pointsByUser.get(b) - pointsByUser.get(a);
            }
        });
    for (let user of sortedPointsByUser)
    {
        console.log(`${user} | ${pointsByUser.get(user)}`);
    }

    console.log('Submissions:');
    const sortedSubmissionsByLanguage = [...Object.keys(submissionsByLanguage)]
        .sort((a, b) =>
        {
            if (submissionsByLanguage[a] == submissionsByLanguage[b])
            {
                // Ascending by language
                return a.localeCompare(b);
            }
            else
            {
                // Descending ny submissions
                return submissionsByLanguage[b] - submissionsByLanguage[a];
            }
        });
    for (let language of sortedSubmissionsByLanguage)
    {
        console.log(`${language} - ${submissionsByLanguage[language]}`);
    }
}

// solve([
//     'Pesho-Java-84',
//     'Gosho-C#-70',
//     'Gosho-C#-84',
//     'Kiro-C#-94',
//     'exam finished'
// ]);

// solve([
//     'Pesho-Java-91',
//     'Gosho-C#-84',
//     'Kiro-JavaScript-90',
//     'Kiro-C#-50',
//     'Kiro-banned',
//     'exam finished'
// ]);